### PR TITLE
[feat][checkpoint] Allow selective state restore on resume

### PR DESCRIPTION
**Pull request by bvolpato**
## Summary

Allow checkpoint weights to seed a new training phase while independently controlling global-step, dataloader, optimizer, and learning-rate scheduler restoration.

Exact resume remains unchanged because every option defaults to `true`.

## Changes

- Add `trainer.resume_load_global_step`, `trainer.resume_load_dataloader_state`, `trainer.resume_load_optimizer_states`, and `trainer.resume_load_lr_scheduler_states`.
- Reset training to step zero when global-step restoration is disabled.
- Skip stale fully-async UID and epoch state for a new training phase.
- Preserve logical epoch progress and epoch-end saves in sync and fully-async trainers when global-step restoration skips the dataloader cursor.
- Bound fully-async producers to the shortened first epoch's remaining groups when the cursor is skipped. Filtered groups return one permit so replacement sampling still works without leaving surplus buffered generation at epoch teardown.
- Pass optimizer and scheduler restore options to policy and critic checkpoint loading.
- Document selective-state resume and separate checkpoint output paths.
- Add focused sync and fully-async regression coverage.

## Testing

- `pytest tests/train/test_trainer.py tests/train/test_config.py tests/train/test_fully_async_trainer.py -q` (`179 passed`)
- Same focused suites on a clean merge with current `upstream/main` at `a6871df` (`183 passed`)
- `pre-commit run --all-files`
- `uv lock --check`
- documentation type check and production build (`66 pages`)
- `git diff --check`

## Megatron dependency

Merge #2149 before using `trainer.resume_load_optimizer_states=false` with Megatron. It refreshes optimizer master parameters from the loaded model; without it, the first optimizer step can overwrite checkpoint weights with stale master values. Default full-state resume is unaffected.

---

> [!NOTE]
> **Medium Risk**
> Changes core checkpoint resume and fully-async epoch/generation bookkeeping; misconfiguration could desync training step vs data or leave stale optimizer state (Megatron caveat noted in PR).
> 
> **Overview**
> Adds four **`trainer.resume_load_*`** flags (global step, dataloader, optimizer, LR scheduler) so runs can load checkpoint **weights** while starting a new phase with fresh step counters, data cursors, or optimizers. Defaults stay **`true`**, so normal resume behavior is unchanged.
> 
> **`load_checkpoints`** honors the flags: optional step reset to 0, conditional **`data.pt`** restore with **`_dataloader_state_restored`**, and optimizer/scheduler restore passed through to policy/critic dispatch. Docs add a YAML example for new-phase resume and recommend a separate **`ckpt_path`**.
> 
> The **sync trainer** limits the first resumed epoch to the logical remainder when global step is kept but the dataloader cursor is not. **Fully-async** skips **`fully_async_state.pt`** when dataloader restore is off, refactors async restore into **`_restore_async_training_state`**, and uses a **generation-group budget** so producers do not overrun a shortened first epoch (including permit release on filtered groups).
> 
> Regression tests cover selective restore, epoch bounding, and async budget behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 897d65f306612766640edfe401412124bccf86c3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/train/trainer.py [hunk 1]
```diff
@@ -1789,13 +1793,17 @@ def load_checkpoints(self) -> Tuple[int, str]:
...
+            logger.warning(
+                f"Global step mismatch: path={checkpoint_global_step}, saved={saved_global_step}. Using path value."
+            )
 
-        # 2. Load dataloader state if available
-        if io.exists(dataloader_state_path):
+        # 2. Load dataloader state if requested and available
+        if not self.cfg.trainer.resume_load_dataloader_state:
+            logger.info("Skipping dataloader state restore; dataloader will start from beginning")
+        elif io.exists(dataloader_state_path):
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/train/trainer.py]: Fixed in 9b0aeb49. Resume now handles a missing train dataloader explicitly and logs an informational skip. Focused resume tests: 6 passed; trainer tests: 16 passed; full pre-commit and docs typecheck passed.

File: skyrl/train/trainer.py [hunk 2]
```diff
@@ -1789,13 +1793,17 @@ def load_checkpoints(self) -> Tuple[int, str]:
...
+        if saved_global_step != checkpoint_global_step:
+            logger.warning(
+                f"Global step mismatch: path={checkpoint_global_step}, saved={saved_global_step}. Using path value."
+            )
 
-        # 2. Load dataloader state if available
-        if io.exists(dataloader_state_path):
+        # 2. Load dataloader state if requested and available
+        if not self.cfg.trainer.resume_load_dataloader_state:
+            logger.info("Skipping dataloader state restore; dataloader will start from beginning")
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/train/trainer.py:1823]: Fixed in 9b0aeb49. Disabling dataloader-state restore now also skips fully async consumed UID, filtered UID, and epoch cursor state while preserving global-step and staleness accounting. Focused resume tests: 6 passed; fully async tests: 12 passed; full pre-commit passed.

File: skyrl/train/trainer.py [hunk 3]
```diff
@@ -1789,13 +1793,19 @@ def load_checkpoints(self) -> Tuple[int, str]:
...
+                f"Global step mismatch: path={checkpoint_global_step}, saved={saved_global_step}. Using path value."
+            )
 
-        # 2. Load dataloader state if available
-        if io.exists(dataloader_state_path):
+        # 2. Load dataloader state if requested and available
+        if self.train_dataloader is None:
+            logger.info("No train dataloader initialized; skipping dataloader state restore")
+        elif not self.cfg.trainer.resume_load_dataloader_state:
+            logger.info("Skipping dataloader state restore; dataloader will start from beginning")
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/train/trainer.py:1823]: Valid. Sync resume now tracks whether the dataloader cursor restored. When global step is kept without that cursor, the first resumed epoch runs only its logical remainder, so checkpoint, HF export, and reference-update boundaries stay aligned. Focused trainer suite: 21 passed.
